src/cbmc/cbmc_parse_options.cpp

```diff
@@ -470,47 +472,93 @@ int cbmc_parse_optionst::doit()
   register_languages();
 
   if(cmdline.isset("test-preprocessor"))
-    return test_c_preprocessor(ui_message_handler)?8:0;
+    return test_c_preprocessor(get_message_handler())?8:0;
 
   if(cmdline.isset("preprocess"))
   {
     preprocessing();
     return 0; // should contemplate EX_OK from sysexits.h
   }
 
-  goto_functionst goto_functions;
+  if(cmdline.isset("show-parse-tree"))
+  {
+    if(cmdline.args.size()!=1 ||
+       is_goto_binary(cmdline.args[0]))
+    {
+      error() << "Please give exactly one source file" << eom;
+      return 6;
+    }
+
+    std::string filename=cmdline.args[0];
+
+    #ifdef _MSC_VER
+    std::ifstream infile(widen(filename));
+    #else
+    std::ifstream infile(filename);
+    #endif
+
+    if(!infile)
+    {
+      error() << "failed to open input file `"
+              << filename << "'" << eom;
+      return 6;
+    }
+
+    std::unique_ptr<languaget> language=
+      get_language_from_filename(filename);
+
+    if(language==nullptr)
+    {
+      error() << "failed to figure out type of file `"
+              <<  filename << "'" << eom;
+      return 6;
+    }
+
+    language->get_language_options(cmdline);
+    language->set_message_handler(get_message_handler());
+
+    status() << "Parsing " << filename << eom;
+
+    if(language->parse(infile, filename))
+    {
+      error() << "PARSING ERROR" << eom;
+      return 6;
+    }
 
-  expr_listt bmc_constraints;
+    language->show_parse(std::cout);
+    return 0;
+  }
 
-  int get_goto_program_ret=
-    get_goto_program(options, bmc_constraints, goto_functions);
+  int get_goto_program_ret=get_goto_program(options);
 
   if(get_goto_program_ret!=-1)
     return get_goto_program_ret;
 
   if(cmdline.isset("show-claims") || // will go away
      cmdline.isset("show-properties")) // use this one
   {
-    const namespacet ns(symbol_table);
-    show_properties(ns, get_ui(), goto_functions);
+    show_properties(goto_model, ui_message_handler.get_ui());
     return 0; // should contemplate EX_OK from sysexits.h
   }
 
-  if(set_properties(goto_functions))
+  if(set_properties())
     return 7; // should contemplate EX_USAGE from sysexits.h
 
   // unwinds <clinit> loops to number of enum elements
   // side effect: add this as explicit unwind to unwind set
   if(options.get_bool_option("java-unwind-enum-static"))
     remove_static_init_loops(
-      symbol_table,
-      goto_functions,
+      goto_model.symbol_table,
+      goto_model.goto_functions,
```

Comment:
The API of `remove_static_init_loops` should be changed to accept a `goto_modelt`. (Which is pretty trivial, because this is the only invocation.)

Reply:
Done